File: LevelOrderTraversal/LevelOrderTraversal.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

//Time Complexity - O(n)
//Space Complexity - O(n)

class LevelOrderTraversal {
	public List<List<Integer>> levelOrder(TreeNode root) {
		
		List<List<Integer>> result = new ArrayList<>();
		
		if(root == null) return result;

		Queue<TreeNode> queue = new LinkedList<>();
		queue.add(root);

		while(!queue.isEmpty()) {
			List<Integer> temp = new ArrayList<>();
			int count = queue.size();

			for(int i = 0; i < count; i++) {

				TreeNode node = queue.poll();
				temp.add(node.val);

				if(node.left != null)
					queue.add(node.left);
				
				if(node.right != null)
					queue.add(node.right);
			}
			result.add(temp);
		}
		return result;
	}
}

File: LevelOrderTraversal.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace BinaryTree
{
    public class LevelOrderTraversal
    {
        /*
         *  T.C: O(N) since traversing all elements of tree
         *  S.C: O(h) height of tree
         */
        List<IList<int>> result = new List<IList<int>>();
        public IList<IList<int>> LevelOrder(TreeNode root)
        {

            dfs(root, 0);
            return result;
        }

        private void dfs(TreeNode root, int level)
        {
            if (root == null)
                return;

            if (result.Count == level)
            {
                result.Add(new List<int>());
            }

            result[level].Add(root.val);

            dfs(root.left, level + 1);
            dfs(root.right, level + 1);
        }



        /*
         * T.C: O(N) traversing through all the elements of tree
         * S.C: O(N) actual space is n/2  but asymptotically N
         */
        public IList<IList<int>> LevelOrderBFS(TreeNode root)
        {
            List<IList<int>> result = new List<IList<int>>();

            Queue<TreeNode> queue = new Queue<TreeNode>();

           

            queue.Enqueue(root);

            while(queue.Count>0)
            {
                int size = queue.Count;
                List<int> lst = new List<int>();
                for (int i = 0; i < size; i++)
                {
                    var element = queue.Dequeue();

                    if(element.left!=null)
                    queue.Enqueue(element.left);

                    if(element.right!=null)
                    queue.Enqueue(element.right);
                   
                    lst.Add(element.val);
                   
                }

                result.Add(lst);
            }

            return result;
        }
    }
